catalog/lease: add support for dual writing session based leases

fqazi opened the pull request:
This patch supports writing session-based leases from the lease manager when the dual_write or higher modes are enabled using an experimental setting. It also supports cleaning orphaned session-based leases on startup, which allows them to be cleaned up. It makes no functional changes to the leasing manager to utilize the session-based lease information, enables the dual write mode for the leasing package, and metamorphically enables dual writes.


The first four commits can be ignored and belong to the following PRs:
https://github.com/cockroachdb/cockroach/pull/111345
https://github.com/cockroachdb/cockroach/pull/111427
https://github.com/cockroachdb/cockroach/pull/111462

informs: https://github.com/cockroachdb/cockroach/issues/95765

fqazi committed:
catalog/lease: add support for dual writing session based leases

Previously, we added support for writing session-based leases to storage
if populated. But outside of unit tests, nothing populated these fields.
This patch supports populating session-based leases when the dual_write
or higher setting is set on the leasing manager. This will cause the
leasing manager to generate entries for expiry-based and session-based
leases. The session based leasing will only be written to storage and
not used. The leasing unit tests run in dual write mode to ensure no
behaviour changes occur.

Epic: CC-24173
Release note: None

fqazi committed:
catalog/lease: cleanup orphaned session based leases

The previous commit added support for writing session-based leases and
cleaning them up. It, however, didn't deal with orphaned ones after a
node crashes. On start-up, these leases should be automatically cleaned
up by the leasing manager. To address this, this patch modifies the
existing query to join against the crdb_internal.kv_session_based_leases
table to extract them and add them to the stored lease.

Epic: CC-24173
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/111658)
<!-- Reviewable:end -->

chrisseto commented on pkg/sql/catalog/lease/descriptor_version_state.go:
```diff
@@ -115,7 +124,18 @@ func (s *descriptorVersionState) hasExpired(timestamp hlc.Timestamp) bool {
 // operating) at the given timestamp.
 func (s *descriptorVersionState) hasExpiredLocked(timestamp hlc.Timestamp) bool {
-	return s.mu.expiration.LessEq(timestamp)
+	// Pick the longest expiration between the session and
```
Very similar comment to the last PR 😅 , place the description of the condition ahead of the action description when the comment is outside of the `if` block.

chrisseto commented on pkg/sql/catalog/lease/descriptor_version_state.go:
```diff
@@ -65,6 +67,9 @@ type descriptorVersionState struct {
 		expiration hlc.Timestamp
 
+		// The session that was used to acquire this descriptor version.
+		session sqlliveness.Session
```
I see a lot of `nil` checks across this struct. When is `session` expected to be nil vs not? Can that be noted on the field itself?

chrisseto commented on pkg/sql/catalog/lease/descriptor_version_state.go:
```diff
@@ -134,7 +154,30 @@ func (s *descriptorVersionState) incRefCountLocked(ctx context.Context, expensiv
 	defer s.mu.Unlock()
-	return s.mu.expiration
+
+	// Pick the longest expiration between the session and
```
Can we utilize a helper in this function and `hasExpiredLocked`? The logic feels tricky enough that having a single definition is worth while.

chrisseto commented on pkg/sql/catalog/lease/lease.go:
```diff
@@ -625,7 +653,9 @@ func releaseLease(ctx context.Context, lease *storedLease, m *Manager) {
 			m.storage.release(ctx, m.stopper, lease)
+			notify()
 		}); err != nil {
+		notify()
```
Should `notify` actually be invoked here if the lease was not released? If so, can you document why?

chrisseto commented on pkg/sql/catalog/lease/lease.go:
```diff
@@ -610,9 +625,22 @@ func acquireNodeLease(
+	releaseLeaseWithNotify(ctx, lease, m, nil)
+}
+
+// releaseLeaseWithNotify from store and notifies a channel.
```
Is "notify" a standard term for closing a channel? notify, to me, reads as "Sends a value".

chrisseto commented on pkg/sql/catalog/lease/storage.go:
```diff
@@ -123,10 +124,15 @@ func (s storage) crossValidateDuringRenewal() bool {
+	minExpiration hlc.Timestamp,
+	session sqlliveness.Session,
+	id descpb.ID,
+	lastLease *storedLease,
```
Is it possible for lastLeast to be `nil` if acquire is being called for the first time?

chrisseto commented on pkg/sql/catalog/lease/storage.go:
```diff
@@ -180,12 +187,25 @@ func (s storage) acquire(
+		}
+		// For session based leases, we will only insert if the version changes
+		// or if the current lease doesn't have a session ID.
+		if lastLease.sessionID != nil &&
```
I'm thoroughly confused about this comment and block of code

Does current lease here mean `lastLease`? If so, shouldn't the check be `lastLeast.sessionID == nil`?

It also says "we will only insert if ..." but `insertLease` is always called. Is `insertLease` doing some extra logic or is there a missing return that will be present for session only mode?

chrisseto commented on pkg/sql/catalog/lease/storage.go:
```diff
@@ -123,10 +124,15 @@ func (s storage) crossValidateDuringRenewal() bool {
-	ctx context.Context, minExpiration hlc.Timestamp, id descpb.ID,
+	ctx context.Context,
+	minExpiration hlc.Timestamp,
+	session sqlliveness.Session,
```
Can you note when it is or is not acceptable for `session` to be `nil`?

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111658)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chrisseto, @ericharmeling, and @michae2)

___
*[`pkg/sql/catalog/lease/descriptor_version_state.go` line 71 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111658#gh-1387098552:-NjpVTH_0KRNnRm8dKlb:b-ynm97k) ([raw file](https://github.com/cockroachdb/cockroach/blob/8937485ab25a6403191df04a64d1e6aaf805a152/pkg/sql/catalog/lease/descriptor_version_state.go#L71)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

I see a lot of `nil` checks across this struct. When is `session` expected to be nil vs not? Can that be noted on the field itself?
</blockquote></details>

Good point:

// The session that was used to acquire this descriptor version, which is  
// only populated when the session based leasing mode is \*at least\* dual  
// write.
___
*[`pkg/sql/catalog/lease/descriptor_version_state.go` line 127 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111658#gh-1387097991:-NjpUxyo69qtkyDhRzUt:bfa20pi) ([raw file](https://github.com/cockroachdb/cockroach/blob/8937485ab25a6403191df04a64d1e6aaf805a152/pkg/sql/catalog/lease/descriptor_version_state.go#L127)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Very similar comment to the last PR 😅 , place the description of the condition ahead of the action description when the comment is outside of the `if` block.
</blockquote></details>


Hopefully this is much clearer:

// A descriptor version state can now potentially contain two different types  
// of expiration:  
// 1) Fixed expirations, which will be based on some timestamp in the future,  
//   that will need to be renewed to keep a descriptor as "active"  
// 2) Session-based expirations, which say that a descriptor is in use,  
//    as long as the sqlliveness exists for it.  
// We are going to pick the longest possible leases between these two options,  
// assuming that session-based leases are being enforced. Session-based leases  
// will only be enforced once the Drain leasing mode is reached, which will stop  
// allowing fixed expiration leases from renewing (i.e. those leases will  
// eventually be \*drained\*.
___
*[`pkg/sql/catalog/lease/descriptor_version_state.go` line 158 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111658#gh-1387100451:-NjpW7-p0PIOGM2gfkOh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8937485ab25a6403191df04a64d1e6aaf805a152/pkg/sql/catalog/lease/descriptor_version_state.go#L158)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Can we utilize a helper in this function and `hasExpiredLocked`? The logic feels tricky enough that having a single definition is worth while.
</blockquote></details>

Done.
___
*[`pkg/sql/catalog/lease/lease.go` line 631 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111658#gh-1387105725:-NjrfqBy8d3Ayz1gPAMg:bbhlfy4) ([raw file](https://github.com/cockroachdb/cockroach/blob/8937485ab25a6403191df04a64d1e6aaf805a152/pkg/sql/catalog/lease/lease.go#L631)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Is "notify" a standard term for closing a channel? notify, to me, reads as "Sends a value".
</blockquote></details>

Done.

I removed all the channel stuff and just simplified this to a new arg to make it async.
___
*[`pkg/sql/catalog/lease/lease.go` line 658 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111658#gh-1387104513:-Njrg22GCqVEWGNLzqp_:b-xogjlk) ([raw file](https://github.com/cockroachdb/cockroach/blob/8937485ab25a6403191df04a64d1e6aaf805a152/pkg/sql/catalog/lease/lease.go#L658)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Should `notify` actually be invoked here if the lease was not released? If so, can you document why?
</blockquote></details>

Done.

Remove this code, but that was a bit suspicious too.
___
*[`pkg/sql/catalog/lease/storage.go` line 129 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111658#gh-1387132822:-NjrWMWnDmsvLdmLFXY0:bqp8vtm) ([raw file](https://github.com/cockroachdb/cockroach/blob/8937485ab25a6403191df04a64d1e6aaf805a152/pkg/sql/catalog/lease/storage.go#L129)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Can you note when it is or is not acceptable for `session` to be `nil`?
</blockquote></details>

Added: 
// A non-nil session should be provided when session based leasing is enabled,  
// which will cause stored leases to populated sessionIDs.

At the caller where the session is fetched I updated the comment:
// A session will be populated within the leasing infrastructure only when  
// session based leasing is enabled. This session will be stored both inside  
// the leases table and descriptor version states in memory, and can be  
// consulted for the expiry depending on the mode // (see SessionBasedLeasingMode).
___
*[`pkg/sql/catalog/lease/storage.go` line 131 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111658#gh-1387116983:-NjrUP9LCxIiguK7B8hE:b-n7zebm) ([raw file](https://github.com/cockroachdb/cockroach/blob/8937485ab25a6403191df04a64d1e6aaf805a152/pkg/sql/catalog/lease/storage.go#L131)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Is it possible for lastLeast to be `nil` if acquire is being called for the first time?
</blockquote></details>

I changed this to add a check and made this entire setup saner. Previously, if it didn't exist we would generate an empty structure.
___
*[`pkg/sql/catalog/lease/storage.go` line 203 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111658#gh-1387132710:-NjrUwM_7a5PorZ6DWY3:bg3n7e5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8937485ab25a6403191df04a64d1e6aaf805a152/pkg/sql/catalog/lease/storage.go#L203)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

I'm thoroughly confused about this comment and block of code

Does current lease here mean `lastLease`? If so, shouldn't the check be `lastLeast.sessionID == nil`?

It also says "we will only insert if ..." but `insertLease` is always called. Is `insertLease` doing some extra logic or is there a missing return that will be present for session only mode?
</blockquote></details>

Yeah this is terribly phrase, this comment should read:

// When the sessionID is set insert lease will upsert a new lease replacing any  
// existing ones. For session based leases we only need to refresh them if the  
// version has changed or if the previous lease was expiry based (i.e. the  
// sessionID was not set).


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111658)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chrisseto, @ericharmeling, @fqazi, and @michae2)

___
*[`pkg/sql/catalog/lease/descriptor_state.go` line 302 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111658#-Njv7a7B8TdUYK4bOyIh:-Njv7a7B8TdUYK4bOyIi:b-tbe0v9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a769f04e847520d651191eb691e138399b490a29/pkg/sql/catalog/lease/descriptor_state.go#L302)):*
> ```Go
> 			// in with the same session ID, and potentially we can end up racing
> 			// with inserts and deletes on the storage side. For real world scenario,
> 			// this never happens because we release only if a new version exists.
> ```

i think we can avoid this special case; see my comment in `releaseLease`
___
*[`pkg/sql/catalog/lease/descriptor_version_state.go` line 115 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111658#-Njv1uBg2nNDZ0UKtKyI:-Njv1uBh0A0UHRVcTYsy:bppkdy1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a769f04e847520d651191eb691e138399b490a29/pkg/sql/catalog/lease/descriptor_version_state.go#L115)):*
> ```Go
> 		sessionID = s.mu.session.ID().String()
> 	}
> 	return redact.Sprintf("%d(%q,%s) ver=%d:%s, refcount=%d", s.GetID(), s.GetName(), s.GetVersion(), s.mu.expiration, sessionID, s.mu.refcount)
> ```

nit: should sessionID be the third parameter?
___
*[`pkg/sql/catalog/lease/lease.go` line 631 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111658#gh-1387105725:-Njv8it51yXsDMAMXKbC:bx8u0ow) ([raw file](https://github.com/cockroachdb/cockroach/blob/8937485ab25a6403191df04a64d1e6aaf805a152/pkg/sql/catalog/lease/lease.go#L631)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Done.

I removed all the channel stuff and just simplified this to a new arg to make it async.
</blockquote></details>

i am pretty sure we can avoid the async flag entirely if we change the condition to:
```
if m.removeOnceDereferenced() {
```


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111658)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chrisseto, @ericharmeling, @michae2, and @rafiss)

___
*[`pkg/sql/catalog/lease/descriptor_state.go` line 302 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111658#-Njv7a7B8TdUYK4bOyIh:-Njxqtos4hKk3qsFz87g:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a769f04e847520d651191eb691e138399b490a29/pkg/sql/catalog/lease/descriptor_state.go#L302)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think we can avoid this special case; see my comment in `releaseLease`
</blockquote></details>

Done.
___
*[`pkg/sql/catalog/lease/descriptor_version_state.go` line 115 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111658#-Njv1uBg2nNDZ0UKtKyI:-Njxqv33DRzjL9Ng5bpz:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/a769f04e847520d651191eb691e138399b490a29/pkg/sql/catalog/lease/descriptor_version_state.go#L115)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: should sessionID be the third parameter?
</blockquote></details>

Done
___
*[`pkg/sql/catalog/lease/lease.go` line 631 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111658#gh-1387105725:-NjxqxCrARY9kLjsq22Z:bvn24ww) ([raw file](https://github.com/cockroachdb/cockroach/blob/8937485ab25a6403191df04a64d1e6aaf805a152/pkg/sql/catalog/lease/lease.go#L631)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i am pretty sure we can avoid the async flag entirely if we change the condition to:
```
if m.removeOnceDereferenced() {
```
</blockquote></details>

Done.

Nice find and simplifies the code!


<!-- Sent from Reviewable.io -->

fqazi committed:
sql/server: adjust logic for DisableSQLServer flag

Previously, we would still run code paths that needed an active SQL
server, even when the server itself was disabled. This would work fine
because leases never required a liveness session, but these code paths
now hang with the move to session-based leases (waiting for the SQL
instance to start). To address this, this patch will skip logic running
internal SQL for tests that disable the SQL server. For real-world
scenarios, there is zero functional change.

Release note: None

fqazi committed:
catalog/lease: update tests to use liveness sessions per-node

Some of the tests for the leasing infrastructure make multiple lease
managers on a single node. Now that we started using liveness, we can't
have all the nodes have same liveness session ID, since it would violate
the PK constrainct. To address this, this patch will give each of these
lease managers in the tests its own liveness provider.

Release note: none

fqazi commented:
@rafiss @chrisseto RFAL

chrisseto reviewed (commented):
I don't really understand the last commit so I'll leave that to @rafiss!

Most everything else looks good, I'm still drawing blanks on the `insertLease` dance. Let me know if it'd be easier to just discuss synchronously!

> Reviewed 1 of 8 files at r3, 2 of 40 files at r5, 18 of 18 files at r7, 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111658#-:-NkzgsrK1Tg7m4-4vYrz:b-pfkvp8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ericharmeling, @fqazi, @michae2, and @rafiss)

___
*[`pkg/server/server_controller_orchestration.go` line 39 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/111658#-NkzgGFKDhoOGbqkzzUt:-NkzgGFKDhoOGbqkzzUu:b-amn8h9) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f2c6902939c64e1713682958d4095c5b0c11150/pkg/server/server_controller_orchestration.go#L39)):*
> ```Go
> 	ctx context.Context, ie isql.Executor, startInitialSecondaryTenants bool,
> ) error {
> 	// If the SQL server is disabled there are no intitial secondary tenants to
> ```

This comment is a bit confusing because the variable name is `startInitialSecondaryTenants`.

Maybe this parameter would be better off as a member on `serverController`?
___
*[`pkg/sql/catalog/lease/descriptor_version_state.go` line 71 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111658#gh-1387098552:-NkzXTKh10Ku8MPNKWGC:bb6zsme) ([raw file](https://github.com/cockroachdb/cockroach/blob/8937485ab25a6403191df04a64d1e6aaf805a152/pkg/sql/catalog/lease/descriptor_version_state.go#L71)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Good point:

// The session that was used to acquire this descriptor version, which is  
// only populated when the session based leasing mode is \*at least\* dual  
// write.
</blockquote></details>

OOC, do you think it would be possible and less error prone to always populate `session`? Assuming it's not prohibitively difficult to plumb in the leasing mode. Ahead of this change there will always be a session, right?

This is just a musing. We're a bit to far down the line to make a large change and I think it would end up being inconsistent which is likely more harmful in the long run!
___
*[`pkg/sql/catalog/lease/lease.go` line 1463 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/111658#-NkzfAIJ97CMHBY_sxK-:-NkzfAIJ97CMHBY_sxK0:bxshu50) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e15ce60bc491fc4590e126a02850601ee9122ea/pkg/sql/catalog/lease/lease.go#L1463)):*
> ```Go
> 		var wg sync.WaitGroup
> 		defer wg.Wait()
> 		log.Errorf(ctx, "MARKED: %d", len(rows))
> ```

Left over log lines?
___
*[`pkg/sql/catalog/lease/lease.go` line 1465 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/111658#-NkzfC2A27UUp7uvBJcB:-NkzfC2A27UUp7uvBJcC:bxshu50) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e15ce60bc491fc4590e126a02850601ee9122ea/pkg/sql/catalog/lease/lease.go#L1465)):*
> ```Go
> 		log.Errorf(ctx, "MARKED: %d", len(rows))
> 		for i := range rows {
> 			log.Errorf(ctx, "MARKED: %v", rows[i])
> ```

Left over log lines?
___
*[`pkg/sql/catalog/lease/storage.go` line 203 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111658#gh-1387132710:-Nkz_qNc4uDz4su83Ko1:b-rsbato) ([raw file](https://github.com/cockroachdb/cockroach/blob/8937485ab25a6403191df04a64d1e6aaf805a152/pkg/sql/catalog/lease/storage.go#L203)):*
That comment change explains the condition much better but I'm still confused about why we're niling out `sessionID`. I looked at `insertLease` but its not providing any hints to me either.

> 		// existing ones. For session based leases we only need to refresh them if the
>		// version has changed or if the previous lease was expiry based (i.e. the
>		// sessionID was not set).

This comment would imply that there should be a return within this if block because there's no need to call `insertLease` if the stated condition is true.

Is there something unique to dualWrite mode that's causing this to be a bit funky?


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111658)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chrisseto, @ericharmeling, @michae2, and @rafiss)

___
*[`pkg/server/server_controller_orchestration.go` line 39 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/111658#-NkzgGFKDhoOGbqkzzUt:-Nl-Nm2C1n0ULAJxqzBL:b-3bd8xm) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f2c6902939c64e1713682958d4095c5b0c11150/pkg/server/server_controller_orchestration.go#L39)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

This comment is a bit confusing because the variable name is `startInitialSecondaryTenants`.

Maybe this parameter would be better off as a member on `serverController`?
</blockquote></details>

Good point. I added disableSQLServer into serverController and used it here.
___
*[`pkg/sql/catalog/lease/descriptor_version_state.go` line 71 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111658#gh-1387098552:-Nl-diY25C3KIaE0xfBA:by2iwh4) ([raw file](https://github.com/cockroachdb/cockroach/blob/8937485ab25a6403191df04a64d1e6aaf805a152/pkg/sql/catalog/lease/descriptor_version_state.go#L71)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

OOC, do you think it would be possible and less error prone to always populate `session`? Assuming it's not prohibitively difficult to plumb in the leasing mode. Ahead of this change there will always be a session, right?

This is just a musing. We're a bit to far down the line to make a large change and I think it would end up being inconsistent which is likely more harmful in the long run!
</blockquote></details>

I think the problem that we have is that its based on a lease by lease basis if the session will be used for expiration. i.e. If someone starts with session based off and runs through a migration,  its possible that we'll have leases that are expiry only until we get far enough. So, we need someway to distinguish the two
___
*[`pkg/sql/catalog/lease/lease.go` line 1463 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/111658#-NkzfAIJ97CMHBY_sxK-:-Nl-OwET27W5Ez2Qc61m:b-ro20c4) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e15ce60bc491fc4590e126a02850601ee9122ea/pkg/sql/catalog/lease/lease.go#L1463)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Left over log lines?
</blockquote></details>

Done.

Ugh, yeah I was debugging something.
___
*[`pkg/sql/catalog/lease/lease.go` line 1465 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/111658#-NkzfC2A27UUp7uvBJcB:-Nl-OzdU3a3npawmXoDM:bs7wumb) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e15ce60bc491fc4590e126a02850601ee9122ea/pkg/sql/catalog/lease/lease.go#L1465)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Left over log lines?
</blockquote></details>

Done.
___
*[`pkg/sql/catalog/lease/storage.go` line 203 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111658#gh-1387132710:-Nl-VlDvAB6qaoY5SQLk:b-rojwqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/8937485ab25a6403191df04a64d1e6aaf805a152/pkg/sql/catalog/lease/storage.go#L203)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

That comment change explains the condition much better but I'm still confused about why we're niling out `sessionID`. I looked at `insertLease` but its not providing any hints to me either.

> 		// existing ones. For session based leases we only need to refresh them if the
>		// version has changed or if the previous lease was expiry based (i.e. the
>		// sessionID was not set).

This comment would imply that there should be a return within this if block because there's no need to call `insertLease` if the stated condition is true.

Is there something unique to dualWrite mode that's causing this to be a bit funky?
</blockquote></details>

Updated and cleaned up the conditions

// Populate the session the ID for the lease if it has been provided (i.e.  
// session based leasing is enabled), since the KV writer below will use it  
// for generating session based leases.  
// In dual write mode if we know there is lease renewal happening and the  
// previous lease was written with a session ID, we will intentionally not  
// set the session ID. This will cause the KV writer to only generate an expiry  
// based lease row, since we already have a valid session based lease from earlier.  
// We do not expect lease renewals to happen at all once session based leasing  
// is fully adopted.


<!-- Sent from Reviewable.io -->

chrisseto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111658#-:-Nl3x9THDat6NkLJsU8q:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ericharmeling, @fqazi, @michae2, and @rafiss)

___
*[`pkg/sql/catalog/lease/storage.go` line 203 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111658#gh-1387132710:-Nl3x68Z8PKnPvIIiDx3:b6u1a37) ([raw file](https://github.com/cockroachdb/cockroach/blob/8937485ab25a6403191df04a64d1e6aaf805a152/pkg/sql/catalog/lease/storage.go#L203)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Updated and cleaned up the conditions

// Populate the session the ID for the lease if it has been provided (i.e.  
// session based leasing is enabled), since the KV writer below will use it  
// for generating session based leases.  
// In dual write mode if we know there is lease renewal happening and the  
// previous lease was written with a session ID, we will intentionally not  
// set the session ID. This will cause the KV writer to only generate an expiry  
// based lease row, since we already have a valid session based lease from earlier.  
// We do not expect lease renewals to happen at all once session based leasing  
// is fully adopted.
</blockquote></details>

:clap: :clap: :clap: 
___
*[`pkg/sql/catalog/lease/storage.go` line 195 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/111658#-Nl3wtQnBAm_-qjtDlMz:-Nl3wtQnBAm_-qjtDlN-:bo30xyw) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b2f869589c830f4663d6b0c9780f7c98c6c1282/pkg/sql/catalog/lease/storage.go#L195)):*
> ```Go
> 		var isLeaseRenewal bool
> 		var lastLeaseWasWrittenWithSessionID bool
> 		// If there was a previous lease then determine if thithiss a renewal and
> ```

nit: s/thithiss/this/


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
i just had small nits and a question about testing. but looks good once those are resolved

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111658#-:-NlxkUU19OL4wT-5ebF4:bpz68bf)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chrisseto, @ericharmeling, @fqazi, and @michae2)

___
*[`pkg/sql/catalog/lease/descriptor_version_state.go` line 115 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111658#-Njv1uBg2nNDZ0UKtKyI:-NlxiKeK9KZzwWivNHOt:b-huy3pw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a769f04e847520d651191eb691e138399b490a29/pkg/sql/catalog/lease/descriptor_version_state.go#L115)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Done
</blockquote></details>

nit: could we use
```
	return redact.Sprintf("%d(%q,%s) ver=%d:%s, refcount=%d", s.GetID(), s.GetName(), redact.SafeString(sessionID), s.GetVersion(), s.mu.expiration, s.mu.refcount)
```

That way the sessionID doesn't get redacted
___
*[`pkg/sql/catalog/lease/descriptor_version_state.go` line 176 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/111658#-NlxinBxApPr5o4xti77:-NlxinBxApPr5o4xti78:blfa1cg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad765b7a04dd9c27da2d54f01ebc0a85a05dee8c/pkg/sql/catalog/lease/descriptor_version_state.go#L176)):*
> ```Go
> 
> // getStoredLease returns a copy of the stored lease.
> func (s *descriptorVersionState) getStoredLease() storedLease {
> ```

nit: could this return a pointer? (but make the copy inside this function still.) only assuming you think it would be easier to use that way
___
*[`pkg/sql/catalog/lease/lease.go` line 1436 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/111658#-Nlxk1eDBqvHyjiJ4vlc:-Nlxk1eDBqvHyjiJ4vld:b1q0na3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2e5a9fc2bc3feb75be86721d77ec914d0f5f705/pkg/sql/catalog/lease/lease.go#L1436)):*
> ```Go
> 
> 		// Read orphaned leases, and join against the internal session
> 		// table in case we have dual written leases.
> ```

do we already have tests that make sure we clean up orphaned leases? do they need to be updated now?


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111658)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chrisseto, @ericharmeling, @michae2, and @rafiss)

___
*[`pkg/sql/catalog/lease/lease.go` line 1436 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/111658#-Nlxk1eDBqvHyjiJ4vlc:-Nm71tRO2Ql-Hc313z4w:bx2xbha) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2e5a9fc2bc3feb75be86721d77ec914d0f5f705/pkg/sql/catalog/lease/lease.go#L1436)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

do we already have tests that make sure we clean up orphaned leases? do they need to be updated now?
</blockquote></details>

Good point, and I updated the test to default to dual write.
___
*[`pkg/sql/catalog/lease/storage.go` line 203 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111658#gh-1387132710:-Nm71yAP3R46ocuqSGHc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8937485ab25a6403191df04a64d1e6aaf805a152/pkg/sql/catalog/lease/storage.go#L203)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

:clap: :clap: :clap:
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
thanks for addressing everything! exciting that this is getting merged 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111658#-:-NoVLCnE9EKOXxm3h6w5:b-lyvgd3)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chrisseto, @ericharmeling, and @michae2)



<!-- Sent from Reviewable.io -->

fqazi commented:
@rafiss TFTR!

fqazi commented:
bors r+
